File: src/main/java/uk/gov/justice/digital/hmpps/riskprofiler/services/NomisService.java
package uk.gov.justice.digital.hmpps.riskprofiler.services;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClientResponseException;
import uk.gov.justice.digital.hmpps.riskprofiler.model.Alert;
import uk.gov.justice.digital.hmpps.riskprofiler.model.BookingDetails;
import uk.gov.justice.digital.hmpps.riskprofiler.model.IncidentCase;
import uk.gov.justice.digital.hmpps.riskprofiler.model.OffenderBooking;
import uk.gov.justice.digital.hmpps.riskprofiler.model.OffenderSentenceTerms;
import uk.gov.justice.digital.hmpps.riskprofiler.model.PagingAndSortingDto;

import javax.validation.constraints.NotNull;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

import static java.lang.String.format;

@Service
@Slf4j
public class NomisService {
    private static final ParameterizedTypeReference<List<Alert>> ALERTS = new ParameterizedTypeReference<>() {
    };
    private static final ParameterizedTypeReference<List<IncidentCase>> INCIDENTS = new ParameterizedTypeReference<>() {
    };
    private static final ParameterizedTypeReference<List<Map>> MAP = new ParameterizedTypeReference<>() {
    };
    private static final ParameterizedTypeReference<List<OffenderBooking>> BOOKING_DETAILS = new ParameterizedTypeReference<>() {
    };
    private static final ParameterizedTypeReference<List<OffenderSentenceTerms>> SENTENCE_TERMS = new ParameterizedTypeReference<>() {
    };
    public static final List<String> ESCAPE_LIST_ALERT_TYPES = List.of("XER", "XEL");

    public static final List<String> SOC_ALERT_TYPES = List.of("PL3", "PVN", "HPI", "XCO", "XD", "XEAN", "XEBM",
            "XFO", "XGANG", "XOCGN", "XP", "XSC");

    private final WebClientCallHelper webClientCallHelper;
    private final List<String> incidentTypes;
    private final List<String> participationRoles;

    public NomisService(final WebClientCallHelper webClientCallHelper,
                        @Value("${app.assaults.incident.types:ASSAULT}") final List<String> incidentTypes,
                        @Value("${app.assaults.participation.roles}") final List<String> participationRoles) {
        this.webClientCallHelper = webClientCallHelper;
        this.incidentTypes = incidentTypes;
        this.participationRoles = participationRoles;
    }

    @Cacheable("escapeAlert")
    public List<Alert> getEscapeListAlertsForOffender(final String nomsId) {
        log.info("Getting escape list alerts for noms id {}", nomsId);
        return getAlertsForOffender(nomsId, ESCAPE_LIST_ALERT_TYPES);
    }

    @CacheEvict("escapeAlert")
    public void evictEscapeListAlertsCache(final String nomsId) {
        log.info("Evicting {} from escapeAlert cache", nomsId);
    }

    @Cacheable("socAlert")
    public List<Alert> getSocListAlertsForOffender(final String nomsId) {
        log.info("Getting soc list alerts for noms id {}", nomsId);
        return getAlertsForOffender(nomsId, SOC_ALERT_TYPES);
    }

    @CacheEvict("socAlert")
    public void evictSocListAlertsCache(final String nomsId) {
        log.info("Evicting {} from socAlert cache", nomsId);
    }

    List<Alert> getAlertsForOffender(final String nomsId, final List<String> alertTypes) {
        log.info("Getting alerts for noms id {} and types {}", nomsId, alertTypes);

        final var types = alertTypes.stream()
                .map(alertType -> format("alertCode:eq:'%s'", alertType))
                .collect(Collectors.joining(",or:"));

        final var uriAlertsForOffenderByType = format("/api/offenders/%1$s/alerts?query=%2$s&latestOnly=false",nomsId, types);
        return webClientCallHelper.getForList(uriAlertsForOffenderByType, ALERTS).getBody();
    }

    public List<OffenderSentenceTerms> getSentencesForOffender(final Long bookingId) {
        log.info("Getting sentences for bookingId {}", bookingId);
        final var uri = format("/api/offender-sentences/booking/%s/sentenceTerms", bookingId);
        return webClientCallHelper.getForList(uri, SENTENCE_TERMS).getBody();
    }

    @Cacheable("incident")
    public List<IncidentCase> getIncidents(@NotNull final String nomsId) {
        log.info("Getting incidents for noms id {} and type {}, with roles of {}", nomsId, incidentTypes, participationRoles);

        final var incidentTypesStr = incidentTypes.stream()
                .map(incidentType -> format("incidentType=%s", incidentType))
                .collect(Collectors.joining("&"));

        final var participationRolesStr = participationRoles.stream()
                .map(participationRole -> format("participationRoles=%s", participationRole))
                .collect(Collectors.joining("&"));

        final var uriIncidentsForOffender = format("/api/offenders/%s/incidents?%s&%s", nomsId, incidentTypesStr, participationRolesStr);
        return webClientCallHelper.getForList(uriIncidentsForOffender, INCIDENTS).getBody();
    }

    @CacheEvict("incident")
    public void evictIncidentsCache(final String nomsId) {
        log.info("Evicting {} from incident cache", nomsId);
    }

    public List<String> getOffendersAtPrison(@NotNull final String prisonId) {
        final var uri = format("/api/bookings?query=agencyId:eq:'%s'", prisonId);
        final List<Map> results;
        try {
            results = webClientCallHelper.getWithPaging(uri, new PagingAndSortingDto(0L, (long) Integer.MAX_VALUE), MAP).getBody();
        } catch (final WebClientResponseException.NotFound e) {
            log.warn("Prison does not exist");
            return List.of();
        }
        return results.stream().map(m -> (String) m.get("offenderNo")).collect(Collectors.toList());
    }

    public List<OffenderBooking> getBookingDetails(final Long bookingId) {
        log.info("Getting details for bookingId {}", bookingId);
        final var uri = // new UriTemplate("/bookings?bookingId={bookingId}").expand(bookingId);
          String.format("/api/bookings?bookingId=%s", bookingId);
        return webClientCallHelper.getForList(uri, BOOKING_DETAILS).getBody();
    }

    public List<String> getMainOffences(final Long bookingId) {
        final var uri = format("/api/bookings/%d/mainOffence", bookingId);
        final var results = webClientCallHelper.getForList(uri, MAP).getBody();
        return results.stream().map(m -> (String) m.get("offenceDescription")).collect(Collectors.toList());
    }

    public String getOffender(@NotNull final Long bookingId) {
        final var uri = format("/api/bookings/%d?basicInfo=true", bookingId);
        final var result = webClientCallHelper.get(uri, BookingDetails.class);
        return result.getOffenderNo();
    }

    public Long getBooking(@NotNull final String nomsId) {
        final var uri = format("/api/bookings/offenderNo/%s", nomsId);
        final var result = webClientCallHelper.get(uri, BookingDetails.class);
        return result.getBookingId();
    }

    public List<String> getPartiesOfIncident(@NotNull final Long incidentId) {
        final var uri = format("/api/incidents/%d", incidentId);
        final IncidentCase incident;
        try {
            incident = webClientCallHelper.get(uri, IncidentCase.class);
        } catch (final WebClientResponseException.NotFound nf) {
            // 404: incident not found, OR has no questions answered yet
            return Collections.emptyList();
        }
        if (incident.getParties() == null || !incidentTypes.contains(incident.getIncidentType())) {
            return Collections.emptyList();
        }
        return incident.getParties().stream()
                .map(party -> party.getBookingId() == null ? null : getOffender(party.getBookingId()))
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/services/NomisService.kt
package uk.gov.justice.digital.hmpps.riskprofiler.services

import jakarta.validation.constraints.NotNull
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.Cacheable
import org.springframework.core.ParameterizedTypeReference
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClientResponseException
import uk.gov.justice.digital.hmpps.riskprofiler.model.Alert
import uk.gov.justice.digital.hmpps.riskprofiler.model.BookingDetails
import uk.gov.justice.digital.hmpps.riskprofiler.model.IncidentCase
import uk.gov.justice.digital.hmpps.riskprofiler.model.OffenderBooking
import uk.gov.justice.digital.hmpps.riskprofiler.model.OffenderSentenceTerms
import java.util.Objects
import java.util.stream.Collectors

@Service
class NomisService(
  private val webClientCallHelper: WebClientCallHelper,
  @param:Value("\${app.assaults.incident.types:ASSAULT}") private val incidentTypes: List<String?>,
  @param:Value("\${app.assaults.participation.roles}") private val participationRoles: List<String>
) {
  @Cacheable("escapeAlert")
  fun getEscapeListAlertsForOffender(nomsId: String): List<Alert> {
    log.info("Getting escape list alerts for noms id {}", nomsId)
    return getAlertsForOffender(nomsId, ESCAPE_LIST_ALERT_TYPES)
  }

  @CacheEvict("escapeAlert")
  fun evictEscapeListAlertsCache(nomsId: String?) {
    log.info("Evicting {} from escapeAlert cache", nomsId)
  }

  @Cacheable("socAlert")
  fun getSocListAlertsForOffender(nomsId: String?): List<Alert> {
    log.info("Getting soc list alerts for noms id {}", nomsId)
    return getAlertsForOffender(nomsId, SOC_ALERT_TYPES)
  }

  @CacheEvict("socAlert")
  fun evictSocListAlertsCache(nomsId: String?) {
    log.info("Evicting {} from socAlert cache", nomsId)
  }

  fun getAlertsForOffender(nomsId: String?, alertCodeList: List<String?>): List<Alert> {
    log.info("Getting alerts for noms id {} and codes {}", nomsId, alertCodeList)
    val alertCodes = alertCodeList.stream()
      .collect(Collectors.joining(","))
    val uriAlertsForOffenderByType =
      String.format("/api/offenders/$nomsId/alerts/v2?alertCodes=$alertCodes")
    return webClientCallHelper.getForList(uriAlertsForOffenderByType, ALERTS).body!!
  }

  fun getSentencesForOffender(bookingId: Long?): List<OffenderSentenceTerms> {
    log.info("Getting sentences for bookingId {}", bookingId)
    val uri = String.format("/api/offender-sentences/booking/%s/sentenceTerms", bookingId)
    return webClientCallHelper.getForList(uri, SENTENCE_TERMS).body!!
  }

  @Cacheable("incident")
  fun getIncidents(nomsId: @NotNull String?): List<IncidentCase> {
    log.info(
      "Getting incidents for noms id {} and type {}, with roles of {}",
      nomsId,
      incidentTypes,
      participationRoles
    )
    val incidentTypesStr = incidentTypes.stream()
      .map { incidentType: String? -> String.format("incidentType=%s", incidentType) }
      .collect(Collectors.joining("&"))
    val participationRolesStr = participationRoles.stream()
      .map { participationRole: String? -> String.format("participationRoles=%s", participationRole) }
      .collect(Collectors.joining("&"))
    val uriIncidentsForOffender =
      String.format("/api/offenders/%s/incidents?%s&%s", nomsId, incidentTypesStr, participationRolesStr)
    return webClientCallHelper.getForList(uriIncidentsForOffender, INCIDENTS).body!!
  }

  @CacheEvict("incident")
  fun evictIncidentsCache(nomsId: String?) {
    log.info("Evicting {} from incident cache", nomsId)
  }

  fun getOffendersAtPrison(prisonId: @NotNull String?): List<String> {
    log.info(String.format("Getting full list of prisoners in prison %s", prisonId))
    val uri = String.format("/api/bookings/v2?prisonId=%s&size=%d", prisonId, 4000)
    val results: RestResponsePage<OffenderBooking> = try {
      webClientCallHelper.getPageRestResponse(uri, CONTENT_MAP).body!!
    } catch (e: WebClientResponseException.NotFound) {
      log.warn("Prison does not exist")
      return listOf()
    }

    val prisonerList = results.content.stream().map { m -> m.offenderNo!! }.collect(Collectors.toList())
    log.debug(String.format("Found %d prisoners currently in %s", prisonerList.size, prisonId))
    return prisonerList
  }

  fun getBookingDetails(bookingId: Long?): List<OffenderBooking> {
    log.info("Getting details for bookingId {}", bookingId)
    val uri = String.format("/api/bookings/v2?bookingId=%s&legalInfo=true", bookingId)
    return webClientCallHelper.getPageRestResponse(uri, BOOKING_DETAILS).body!!.content
  }

  fun getMainOffences(bookingId: Long?): List<String> {
    val uri = String.format("/api/bookings/%d/mainOffence", bookingId)
    val results = webClientCallHelper.getForList(uri, MAP).body
    return results!!.stream().map { m: Map<*, *> -> m["offenceDescription"] as String }
      .collect(Collectors.toList())
  }

  fun getOffender(bookingId: @NotNull Long?): String? {
    val uri = String.format("/api/bookings/%d?basicInfo=true", bookingId)
    val result = webClientCallHelper.get(uri, BookingDetails::class.java)
    return result.offenderNo
  }

  fun getBooking(nomsId: @NotNull String?): Long {
    val uri = String.format("/api/bookings/offenderNo/%s", nomsId)
    val (bookingId) = webClientCallHelper.get(uri, BookingDetails::class.java)
    return bookingId!!
  }

  fun getPartiesOfIncident(incidentId: @NotNull Long?): List<String?> {
    val uri = String.format("/api/incidents/%d", incidentId)
    val incident: IncidentCase
    incident = try {
      webClientCallHelper.get(uri, IncidentCase::class.java)
    } catch (nf: WebClientResponseException.NotFound) {
      // 404: incident not found, OR has no questions answered yet
      return emptyList<String>()
    }
    return if (incident.parties == null || !incidentTypes.contains(incident.incidentType)) {
      emptyList<String>()
    } else incident.parties!!.stream()
      .map { (bookingId) -> bookingId?.let { getOffender(it) } }
      .filter { obj: String? -> Objects.nonNull(obj) }
      .collect(Collectors.toList())
  }

  companion object {
    private val log = LoggerFactory.getLogger(NomisService::class.java)
    private val ALERTS: ParameterizedTypeReference<List<Alert>> = object : ParameterizedTypeReference<List<Alert>>() {}
    private val INCIDENTS: ParameterizedTypeReference<List<IncidentCase>> =
      object : ParameterizedTypeReference<List<IncidentCase>>() {}
    private val MAP: ParameterizedTypeReference<List<Map<*, *>>> =
      object : ParameterizedTypeReference<List<Map<*, *>>>() {}
    private val CONTENT_MAP: ParameterizedTypeReference<RestResponsePage<OffenderBooking>> =
      object : ParameterizedTypeReference<RestResponsePage<OffenderBooking>>() {}
    private val BOOKING_DETAILS: ParameterizedTypeReference<RestResponsePage<OffenderBooking>> =
      object : ParameterizedTypeReference<RestResponsePage<OffenderBooking>>() {}
    private val SENTENCE_TERMS: ParameterizedTypeReference<List<OffenderSentenceTerms>> =
      object : ParameterizedTypeReference<List<OffenderSentenceTerms>>() {}
    val ESCAPE_LIST_ALERT_TYPES = java.util.List.of("XER", "XEL")
    val SOC_ALERT_TYPES = java.util.List.of(
      "PL3", "PVN", "HPI", "XCO", "XD", "XEAN", "XEBM",
      "XFO", "XGANG", "XOCGN", "XP", "XSC"
    )
  }
}
